rdtools/availability.py

```diff
@@ -0,0 +1,576 @@
+            production_fill *= row['energy_actual'] / subset_energy.sum()
+            corrected_segment = row['energy_start'] + production_fill
+            corrected_cumulative_energy.loc[start:end] = corrected_segment
+
+        self.energy_cumulative_corrected = corrected_cumulative_energy
+        self.loss_system = lost_power_full
+
+    def _combine_losses(self, rollup_period='M'):
+        """
+        Combine subsystem and system losses.
+
+        Sets the `loss_total` and `results` attributes.
+
+        Parameters
+        ----------
+        rollup_period : pandas offset string, default 'M'
+            The period on which to roll up losses and calculate availability.
+        """
+        self.loss_total = self.loss_system + self.loss_subsystem
+
+        # calculate actual production based on corrected cumulative meter
+        cumulative_energy = self.energy_cumulative_corrected
+        resampled_cumulative = cumulative_energy.resample(rollup_period)
+        actual_production = (
+            resampled_cumulative.last() - resampled_cumulative.first()
+        )
+
+        lost_production = rdtools.energy_from_power(self.loss_total)
+        df = pd.DataFrame({
+            'lost_production': lost_production.resample(rollup_period).sum(),
+            'actual_production': actual_production,
+        })
+        loss_plus_actual = df['lost_production'] + df['actual_production']
+        df['availability'] = 1 - df['lost_production'] / loss_plus_actual
+        self.results = df.tz_localize(None)
+
+    def run(self, low_threshold=None, relative_sizes=None,
+            power_system_limit=None, quantiles=(0.01, 0.99),
+            rollup_period='M'):
+        """
+        Run the availability analysis.
+
+        Parameters
+        ----------
+        low_threshold : float or pd.Series, optional
+            An optional threshold used to naively classify subsystems as
+            online. If the threshold is a scalar, it will be used for all
+            subsystems. For subsystems with different capacities, a pandas
+            Series may be passed with index values matching the columns in
+            ``power_subsystem``. Units must match ``power_subsystem`` and
+            ``power_system``. If omitted, the limit is calculated for each
+            subsystem independently as 0.001 times the 99th percentile of its
+            power data.
+
+        relative_sizes : dict or pd.Series, optional
+            The production capacity of each subsystem, normalized by the mean
+            subsystem capacity. If not specified, it will be estimated from
+            power data.
+
+        power_system_limit : float or pd.Series, optional
+            An optional maximum system power used as an upper limit for
+            (power_system + lost_power) so that the maximum system capacity or
+            interconnection limit is not exceeded. If omitted, that check is
+            skipped.
```

Comment:
This could use a little more explanation in the new OO context. In particular it isn't immediately clear what `lost_power` is and that this check is in the subsystem part of the routine. On a related note, would it make sense to add this check to the system-level part of the analysis?

Reply:
The missing context here is large systems with plant controllers and dynamic inverter setpoints.  For example, a system with an interconnection limit of 20MW might have 14x 1.5MW inverters for a total of 14 * 1.5 = 21MW total capacity, but the inverters are normally operated with a clipping point of 20/14=1.428MW so that the system doesn't go over its 20MW limit.  But if one inverter trips offline, the other inverters are allowed to go all the way up to their hard limit of 1.5MW, meaning the system produces 13*1.5=19.5MW in full sun, so really only 0.5MW is lost instead of 1.5 (or 1.428) and naively estimating loss at 1.5 or 1.428 MW would be incorrect.

With that out of the way, I agree that the parameter description needs improving.  Do you think it's worthwhile to describe the use case like the above paragraph?  Or should we just keep it simple and be clearer about what it does. 

> would it make sense to add this check to the system-level part of the analysis?

I don't think the system-level method is really susceptible to the error that this check tries to correct.  The time series system-level power loss (`loss_system`) is created from a scalar loss that is distributed in proportion to the time series expected power signal because that's really the only guidance we have on how to distribute the overall production loss across the outage duration.  So while the system-level loss time series might sometimes exceed the system capacity, that isn't indicative of an error in the loss estimate, just an error of how the loss estimate got distributed across the outage.  I do have a way of distributing production while obeying a maximum capacity, but the code wasn't very elegant and it wasn't always numerically stable, so I didn't pursue it here.

Reply:
Thanks that helps me understand. Let's work through the system-level question before deciding on new docs for it. On that system-level point, I see how the original intent of that parameter isn't really relevant. But, should the same parameter have influence? For example, should it be used to clip power in the calculation of `energy_expected_rescaled`?

Reply:
Hmm, interesting point that the clipping point isn't preserved when rescaling the expected energy to match total actual production.  As long as the expected energy is reasonably accurate (a handful of percent error), it wouldn't be a big error.  If we did make the rescaling operation preserve the system power limit, I think we would need to use some sort of iterative convergent method to still have the property that `energy_expected_rescaled` matches actual production in aggregate during normal production times, which is what I was hoping to avoid.  I would rather just document its limitations and encourage the user to pass in a better expected energy signal if it's a concern. 

Do you see another way to make use of it?

Reply:
> Maximum allowable system power. This parameter is used to account for cases where online subsystems can partially mitigate the loss of an offline subsystem, for example a system with a plant controller and dynamic inverter setpoints.  This constraint is only applied to the subsystem loss calculation.